### ISSUE TITLE
Use protobuf for serialization/deserialization

### ISSUE THREAD

**Issue by jvican** [sbt/zinc#268]
Serialization and deserialization of Analysis files is important for:
  
* Performance. The faster, the better.
* Disk usage.
* Correctness. We want serialised formats to be read by Zinc even if they were done years and years ago. This is super important if we're going to ship analysis files and cache incremental compilation, a project @romanowski is working on in [Hoarder](https://github.com/romanowski/hoarder).
  
The serialisation that we're using now is far than ideal: it's relatively slow, it consumes a lot of space and it does not ensure binary compatibility of serialised files.
  
Because of that, I'd like to propose to use Protobuf. It addresses the three problems at the same time:
  
* It's [the fastest serialisation library out there in the Scala community](https://medium.com/@dkomanov/scala-serialization-419d175c888a#.exqx4eeyf).
* Disk usage is diminished because it outputs binaries. These binaries are small so we can then decide whether we can avoid the compressing/decompressing step based on the resulting sizes, which may bring is some speedup.
* Binary compatibility is taken seriously. The whole protocol is architected around it, so this has the plus benefit that we won't ever break the binaries if serious changes to the API happen.

**Comment by romanowski** [comment]
Our Analysis holds 2 things:
* incremental compiler information
* apis

Latter is mostly used in SBT and I feel that those APIs soon become obsolete (can be replaced with e.g. scala.meta, but I don't want to go into details here).

So we should mostly focus on first part. IMO we need to first migrate parts that use java serialization 
(most can be found [here](https://github.com/sbt/zinc/blob/1.0/internal/zinc-persist/src/main/scala/sbt/internal/inc/TextAnalysisFormat.scala#L34-L48), excluding sbinary ones).

The only concern I have about Protobufs is that debugging may be harder then zipped text files but a) this not apply for thing that are serialized using sbinary/java serialization b) this is only zinc developers (and there is probably a few who reads Analysis manually).

**Comment by jvican** [comment]
No matter how Analysis files are used, they should stay for now always binary compatible. We cannot use the semantic database now, so it's better not to talk about that since it's a decision that needs to be taken at least one year in the future when everything stabilises.
  
> The only concern I have about Protobufs is that debugging may be harder then zipped text files but a) this not apply for thing that are serialized using sbinary/java serialization b) this is only zinc developers (and there is probably a few who reads Analysis manually).
  
I've thought about this. This is not a problem, though, protobufs are so easy that we can just provide readers of analysis files to people interested in having a look at the contents of analysis files. Protobufs do not prohibit debugging by design :wink:.

**Comment by eed3si9n** [comment]
My current plan is to gut out sbinary and replace it with Contraband + sjson-new.
It's meant to be binary compatible for both JSON and the Scala code it generates.
https://github.com/sbt/zinc/issues/203

**Comment by jvican** [comment]
@eed3si9n I'm glad to use Contraband since as you mention it's binary compatible. The only issue with it is that it uses JSON.
  
Choosing JSON over our current binary format will kill the performance of Zinc.
  
These are the problems with JSON:
  
* Serialisation and deserialization is slow.
* Disk usage is multiplied by a big factor, which will affect users in two aspects:
  * Runtime slowdown since compression will take more time.
  * Big disk usage footprint.
  
Right now, we're dealing with analysis files that take up 2GB *in a binary and uncompressed format*. Choosing JSON will increase that size up to a level in which it may be unbearable to use incremental compilation in computers with limited resources.
  
However, my main issue with your proposal is speed. We've been pouring lots of resources to make Zinc faster, and I believe this will make it considerably slower, making our previous efforts insignificant. Remember that serialization/deserialization + compressing/decompressing will happen in every compilation cycle. We probably don't want Scala developers to experience the slowdown that this change would suppose.
  
I understand your effort to unify the dependencies accross sbt modules, but Zinc does not need to represent a network protocol as sbt does, it needs to store data fast. It fundamentally requires another solution.
    
In any case, I still propose to use protobuf, a library that has been heavily battle-tested in all Google codebases and for which speed is a feature. I repeat the motivation:
  
* It's faster.
* It takes up much less disk usage.
* Compressing/decompressing will take less time.
  
That being said, if you really want to use Contraband over protobuf, I'd like to ask you to two things:
1. Explain why you prefer it.
2. Add a binary codec that stores data in a moderately efficient manner so that at least we're not slower than we are right now.
  
To make it more explicit, I *strongly disagree* with the use of Contraband + sjson-new.

**Comment by eed3si9n** [comment]
As a general statement, for an involved discussion like this, we might perhaps pick some other medium other than posting comments. Our thinking has a lot of common ground, but these comments tend to become more contentious.

Serialization problem arises in many different ways in sbt code base. Previous and current sbt teams have been looking to replace sbinary and unify to something else. After meandering around Scala Pickling, we've come around to Contraband + sjson-new. Similar to Pickling, sjson-new is actually able to pick arbitrary backend (different JSON ASTs or even binary formats).

First, we need to get the correctness and machine-independence right. That means avoiding weird automatic macros and runtime serialization behaviors. From that perspective, Google Protobuf is a good contender.

I also agree performance is a key feature. We are both on the same page here. I've looked into benchmarking JSON against binary formats like my own and MessagePack, but the number seems to indicate that straight JSON or gzipped JSON is the fastest. See [sjson-new and the prisoner of Azkaban](http://eed3si9n.com/sjson-new-and-the-prisoner-of-azkaban). I think @fommil has also looked into performance on serialization as well, and has told me gzipped JSON is good enough.

Set aside the performance, Contraband-generated pseudo case class is intended to be evolvable in a binary compatible way. That's relevant wherever the datatype is exposed out to (public) Zinc API, but it might not matter much in other cases where the serialization is an implementation detail.

To wrap up, I'm open to binary or any other backend, but I'd like to see evidence that it in fact is faster than gzipped JSON, and that it matters to overall perf. We can schedule a hangout to discuss more on this. I also echo Krzysztof's concern about debugging.

**Comment by fommil** [comment]
My analysis was of compressed json via spray-json (and latterly spray-json-shapeless) with formats cached in vals. And no Either fields. Plus I sent some PRs around that time to spray-json with perf fixes.

I wouldn't make the assumption that a new json lib will have the same performance characteristics as something that's been battle hardened.

The compression is critical for larger payloads. Ideally this should be done on a separate thread, e.g. see FastJar in sbt-big-project (which is a drop in replacement for Package when sbt moves to java 7+).

I don't think we need protocol buffers, but we definitely need compression, and I can't recommend a new JSON solution, especially not something based on pickling or runtime reflection.

**Comment by fommil** [comment]
Btw, compression helps even on SSD for huge payloads. HDD vs SSD shouldn't matter on linux unless the RAM is contented / synced (kernel caching is pretty good). Large payloads start to force writing to disk.

**Comment by olafurpg** [comment]
> RE: debugging protobuf binary messages

Protobufs can be pretty-printed in a readable text format, see live demo http://thesamet.github.io/scalapbjs-test/ A binary file can be pretty printed like that with the `protoc` command line tool. You can leave out the schema when calling `protoc` and field names are then replaced by their tag identifier, producing output like:
```
1: "Our menu"
2 {
  1: "Apple"
  2: "1 unit (3\" dia)"
  3: 95
}
```

Given the focus on compiler performance in 2.13, I think performance of zinc should be treated of equal importance. Only zinc maintainers have to debug serialized files while every zinc user benefits from faster compilation and less disk usage. The benchmarks [here](https://dkomanov.github.io/scala-serialization/) show that protobuf have clear advantage over Jackson JSON, which is to my best knowledge one of the most performant JSON library out there. The only test where protobuf and compressed JSON perform equally well is file size for small objects (<2kb).

ScalaPB generates idiomatic Scala case classes. We are using ScalaPB [in scala.meta](https://github.com/scalameta/scalameta/pull/748) and my (limited) experience has so far been great.

**Comment by stuhood** [comment]
Pants would be fine with this, as long as the "analysis relocatability" problem (#218) is addressed at the same time. We currently parse analysis in order to rewrite it to be relocatable/cachable/portable, and losing the ability to do that would be a showstopper.

**Comment by ijuma** [comment]
With regards to compression, I'd suggest considering Zstandard. The following article describes the advantages over gzip:

https://code.facebook.com/posts/1658392934479273/smaller-and-faster-data-compression-with-zstandard/

LZ4 is also a good option if speed is more important than compression ratio. The following table from the article above provides a comparison for the Silesia corpus:

![image](https://cloud.githubusercontent.com/assets/24747/26587054/1d62cc64-4549-11e7-9268-1acd7d2fefd1.png)

**Comment by olafurpg** [comment]
Fixed in #351 ?

**Comment by jvican** [comment]
Yeah, closing. I implemented this in #351. Thanks to @tayama0324 for the great insights on why protobuf is more compatible than JSON. For those wishing to know why, please visit the pull request and read the conversation.
